### ISSUE TITLE
Bug in negating logical TRUE

### ISSUE THREAD

**Issue by oroppas** [flang-compiler/f18-llvm-project#967]
`flang` and `gfortran` disagrees with the following code:

```fortran
program main

  logical x

  x = .TRUE.
  print *, "x = ", x
  print *, ".NOT.x"
  print *, .NOT.x

end
```

### flang
```
 x =  T
 .NOT.x
 T
```

### gfortran
```
 x =  T
 .NOT.x
 F
```

**Comment by jeanPerier** [comment]
I cannot reproduce the issue locally. I am getting the expected result with flang and valgrind does not raises any alarms:

```
valgrind ./a.out 
==4037731== Memcheck, a memory error detector
==4037731== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==4037731== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==4037731== Command: ./a.out
==4037731== 
 x =  T
 .NOT.x
 F
==4037731== 
==4037731== HEAP SUMMARY:
==4037731==     in use at exit: 0 bytes in 0 blocks
==4037731==   total heap usage: 5 allocs, 5 frees, 148,992 bytes allocated
==4037731== 
==4037731== All heap blocks were freed -- no leaks are possible
==4037731== 
==4037731== For lists of detected and suppressed errors, rerun with: -s
==4037731== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Can you provide the FIR, LLVM IR, and assembly for this program on your side ?

**Comment by oroppas** [comment]
This is how I compiled

```
bbc -o - ../negate_logical_true.f90 | tco | llc --relocation-model=pic --filetype=obj -o negate_logical_true.o

clang ./negate_logical_true.o -L$HOME/packages/llvm/flang/build/lib -Wl,-rpath -Wl,$HOME/packages/llvm/flang/build/lib -lFortran_main -lFortranRuntime -lFortranDecimal
```

and generate FIR, LLVM IR, and assembly

```
$HOME/bin/bbc negate_logical_true.f90
$HOME/bin/tco negate_logical_true.mlir -o negate_logical_true.ll
$HOME/bin/llc negate_logical_true.ll --relocation-model=pic -o negate_logical_true.s
```

## FIR

```
func @_QQmain() {
  %c6_i32 = constant 6 : i32
  %c4 = constant 4 : index
  %c7_i32 = constant 7 : i32
  %c6 = constant 6 : index
  %c-1_i32 = constant -1 : i32
  %c8_i32 = constant 8 : i32
  %true = constant true
  %0 = fir.alloca !fir.logical<4> {bindc_name = "x", uniq_name = "_QEx"}
  %1 = fir.convert %true : (i1) -> !fir.logical<4>
  fir.store %1 to %0 : !fir.ref<!fir.logical<4>>
  %2 = fir.address_of(@_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000) : !fir.ref<!fir.char<1,29>>
  %3 = fir.convert %2 : (!fir.ref<!fir.char<1,29>>) -> !fir.ref<i8>
  %4 = fir.call @_FortranAioBeginExternalListOutput(%c-1_i32, %3, %c6_i32) : (i32, !fir.ref<i8>, i32) -> !fir.ref<i8>
  %5 = fir.address_of(@_QQcl.78203D20) : !fir.ref<!fir.char<1,4>>
  %6 = fir.convert %5 : (!fir.ref<!fir.char<1,4>>) -> !fir.ref<i8>
  %7 = fir.convert %c4 : (index) -> i64
  %8 = fir.call @_FortranAioOutputAscii(%4, %6, %7) : (!fir.ref<i8>, !fir.ref<i8>, i64) -> i1
  %9 = fir.load %0 : !fir.ref<!fir.logical<4>>
  %10 = fir.convert %9 : (!fir.logical<4>) -> i1
  %11 = fir.call @_FortranAioOutputLogical(%4, %10) : (!fir.ref<i8>, i1) -> i1
  %12 = fir.call @_FortranAioEndIoStatement(%4) : (!fir.ref<i8>) -> i32
  %13 = fir.call @_FortranAioBeginExternalListOutput(%c-1_i32, %3, %c7_i32) : (i32, !fir.ref<i8>, i32) -> !fir.ref<i8>
  %14 = fir.address_of(@_QQcl.2E4E4F542E78) : !fir.ref<!fir.char<1,6>>
  %15 = fir.convert %14 : (!fir.ref<!fir.char<1,6>>) -> !fir.ref<i8>
  %16 = fir.convert %c6 : (index) -> i64
  %17 = fir.call @_FortranAioOutputAscii(%13, %15, %16) : (!fir.ref<i8>, !fir.ref<i8>, i64) -> i1
  %18 = fir.call @_FortranAioEndIoStatement(%13) : (!fir.ref<i8>) -> i32
  %19 = fir.call @_FortranAioBeginExternalListOutput(%c-1_i32, %3, %c8_i32) : (i32, !fir.ref<i8>, i32) -> !fir.ref<i8>
  %20 = fir.load %0 : !fir.ref<!fir.logical<4>>
  %21 = fir.convert %20 : (!fir.logical<4>) -> i1
  %22 = xor %21, %true : i1
  %23 = fir.call @_FortranAioOutputLogical(%19, %22) : (!fir.ref<i8>, i1) -> i1
  %24 = fir.call @_FortranAioEndIoStatement(%19) : (!fir.ref<i8>) -> i32
  return
}
func private @_FortranAioBeginExternalListOutput(i32, !fir.ref<i8>, i32) -> !fir.ref<i8> attributes {fir.io, fir.runtime}
fir.global linkonce @_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000 constant : !fir.char<1,29> {
  %0 = fir.string_lit "./../negate_logical_true.f90\00"(29) : !fir.char<1,29>
  fir.has_value %0 : !fir.char<1,29>
}
func private @_FortranAioOutputAscii(!fir.ref<i8>, !fir.ref<i8>, i64) -> i1 attributes {fir.io, fir.runtime}
fir.global linkonce @_QQcl.78203D20 constant : !fir.char<1,4> {
  %0 = fir.string_lit "x = "(4) : !fir.char<1,4>
  fir.has_value %0 : !fir.char<1,4>
}
func private @_FortranAioOutputLogical(!fir.ref<i8>, i1) -> i1 attributes {fir.io, fir.runtime}
func private @_FortranAioEndIoStatement(!fir.ref<i8>) -> i32 attributes {fir.io, fir.runtime}
fir.global linkonce @_QQcl.2E4E4F542E78 constant : !fir.char<1,6> {
  %0 = fir.string_lit ".NOT.x"(6) : !fir.char<1,6>
  fir.has_value %0 : !fir.char<1,6>
}
```

## LLVM IR

```
; ModuleID = 'FIRModule'
source_filename = "FIRModule"

@_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000 = linkonce constant [29 x i8] c"./../negate_logical_true.f90\00"
@_QQcl.78203D20 = linkonce constant [4 x i8] c"x = "
@_QQcl.2E4E4F542E78 = linkonce constant [6 x i8] c".NOT.x"

declare i8* @malloc(i64)

declare void @free(i8*)

define void @_QQmain() !dbg !3 {
  %1 = alloca i32, i64 1, align 4, !dbg !7
  store i32 -1, i32* %1, align 4, !dbg !9
  %2 = call i8* @_FortranAioBeginExternalListOutput(i32 -1, i8* getelementptr inbounds ([29 x i8], [29 x i8]* @_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000, i32 0, i32 0), i32 6), !dbg !10
  %3 = call i1 @_FortranAioOutputAscii(i8* %2, i8* getelementptr inbounds ([4 x i8], [4 x i8]* @_QQcl.78203D20, i32 0, i32 0), i64 4), !dbg !11
  %4 = load i32, i32* %1, align 4, !dbg !12
  %5 = trunc i32 %4 to i1, !dbg !13
  %6 = call i1 @_FortranAioOutputLogical(i8* %2, i1 %5), !dbg !14
  %7 = call i32 @_FortranAioEndIoStatement(i8* %2), !dbg !15
  %8 = call i8* @_FortranAioBeginExternalListOutput(i32 -1, i8* getelementptr inbounds ([29 x i8], [29 x i8]* @_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000, i32 0, i32 0), i32 7), !dbg !16
  %9 = call i1 @_FortranAioOutputAscii(i8* %8, i8* getelementptr inbounds ([6 x i8], [6 x i8]* @_QQcl.2E4E4F542E78, i32 0, i32 0), i64 6), !dbg !17
  %10 = call i32 @_FortranAioEndIoStatement(i8* %8), !dbg !18
  %11 = call i8* @_FortranAioBeginExternalListOutput(i32 -1, i8* getelementptr inbounds ([29 x i8], [29 x i8]* @_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000, i32 0, i32 0), i32 8), !dbg !19
  %12 = load i32, i32* %1, align 4, !dbg !20
  %13 = trunc i32 %12 to i1, !dbg !21
  %14 = xor i1 %13, true, !dbg !22
  %15 = call i1 @_FortranAioOutputLogical(i8* %11, i1 %14), !dbg !23
  %16 = call i32 @_FortranAioEndIoStatement(i8* %11), !dbg !24
  ret void, !dbg !25
}

declare i8* @_FortranAioBeginExternalListOutput(i32, i8*, i32)

declare i1 @_FortranAioOutputAscii(i8*, i8*, i64)

declare i1 @_FortranAioOutputLogical(i8*, i1)

declare i32 @_FortranAioEndIoStatement(i8*)

!llvm.dbg.cu = !{!0}
!llvm.module.flags = !{!2}

!0 = distinct !DICompileUnit(language: DW_LANG_C, file: !1, producer: "mlir", isOptimized: true, runtimeVersion: 0, emissionKind: FullDebug)
!1 = !DIFile(filename: "FIRModule", directory: "/")
!2 = !{i32 2, !"Debug Info Version", i32 3}
!3 = distinct !DISubprogram(name: "_QQmain", linkageName: "_QQmain", scope: null, file: !4, line: 1, type: !5, scopeLine: 1, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, retainedNodes: !6)
!4 = !DIFile(filename: "negate_logical_true.mlir", directory: "/tmp/negate_logical_true/wrong")
!5 = !DISubroutineType(types: !6)
!6 = !{}
!7 = !DILocation(line: 9, column: 8, scope: !8)
!8 = !DILexicalBlockFile(scope: !3, file: !4, discriminator: 0)
!9 = !DILocation(line: 11, column: 3, scope: !8)
!10 = !DILocation(line: 14, column: 8, scope: !8)
!11 = !DILocation(line: 18, column: 8, scope: !8)
!12 = !DILocation(line: 19, column: 8, scope: !8)
!13 = !DILocation(line: 20, column: 9, scope: !8)
!14 = !DILocation(line: 21, column: 9, scope: !8)
!15 = !DILocation(line: 22, column: 9, scope: !8)
!16 = !DILocation(line: 23, column: 9, scope: !8)
!17 = !DILocation(line: 27, column: 9, scope: !8)
!18 = !DILocation(line: 28, column: 9, scope: !8)
!19 = !DILocation(line: 29, column: 9, scope: !8)
!20 = !DILocation(line: 30, column: 9, scope: !8)
!21 = !DILocation(line: 31, column: 9, scope: !8)
!22 = !DILocation(line: 32, column: 9, scope: !8)
!23 = !DILocation(line: 33, column: 9, scope: !8)
!24 = !DILocation(line: 34, column: 9, scope: !8)
!25 = !DILocation(line: 35, column: 3, scope: !8)
```

## Assembly

```
	.text
	.file	"FIRModule"
	.globl	_QQmain                         # -- Begin function _QQmain
	.p2align	4, 0x90
	.type	_QQmain,@function
_QQmain:                                # @_QQmain
.Lfunc_begin0:
	.file	1 "/tmp/negate_logical_true/wrong/negate_logical_true.mlir"
	.loc	1 1 0                           # negate_logical_true.mlir:1:0
	.cfi_startproc
# %bb.0:
	pushq	%r14
	.cfi_def_cfa_offset 16
	pushq	%rbx
	.cfi_def_cfa_offset 24
	pushq	%rax
	.cfi_def_cfa_offset 32
	.cfi_offset %rbx, -24
	.cfi_offset %r14, -16
	.loc	1 11 3 prologue_end             # negate_logical_true.mlir:11:3
	movl	$-1, 4(%rsp)
	.loc	1 14 8                          # negate_logical_true.mlir:14:8
	movq	_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000@GOTPCREL(%rip), %r14
	movl	$-1, %edi
	movq	%r14, %rsi
	movl	$6, %edx
	callq	_FortranAioBeginExternalListOutput@PLT
	movq	%rax, %rbx
	.loc	1 18 8                          # negate_logical_true.mlir:18:8
	movq	_QQcl.78203D20@GOTPCREL(%rip), %rsi
	movl	$4, %edx
	movq	%rax, %rdi
	callq	_FortranAioOutputAscii@PLT
	.loc	1 19 8                          # negate_logical_true.mlir:19:8
	movl	4(%rsp), %esi
	.loc	1 21 9                          # negate_logical_true.mlir:21:9
	movq	%rbx, %rdi
	callq	_FortranAioOutputLogical@PLT
	.loc	1 22 9                          # negate_logical_true.mlir:22:9
	movq	%rbx, %rdi
	callq	_FortranAioEndIoStatement@PLT
	.loc	1 23 9                          # negate_logical_true.mlir:23:9
	movl	$-1, %edi
	movq	%r14, %rsi
	movl	$7, %edx
	callq	_FortranAioBeginExternalListOutput@PLT
	movq	%rax, %rbx
	.loc	1 27 9                          # negate_logical_true.mlir:27:9
	movq	_QQcl.2E4E4F542E78@GOTPCREL(%rip), %rsi
	movl	$6, %edx
	movq	%rax, %rdi
	callq	_FortranAioOutputAscii@PLT
	.loc	1 28 9                          # negate_logical_true.mlir:28:9
	movq	%rbx, %rdi
	callq	_FortranAioEndIoStatement@PLT
	.loc	1 29 9                          # negate_logical_true.mlir:29:9
	movl	$-1, %edi
	movq	%r14, %rsi
	movl	$8, %edx
	callq	_FortranAioBeginExternalListOutput@PLT
	movq	%rax, %rbx
	.loc	1 30 9                          # negate_logical_true.mlir:30:9
	movb	4(%rsp), %al
	.loc	1 32 9                          # negate_logical_true.mlir:32:9
	xorb	$1, %al
	.loc	1 33 9                          # negate_logical_true.mlir:33:9
	movzbl	%al, %esi
	movq	%rbx, %rdi
	callq	_FortranAioOutputLogical@PLT
	.loc	1 34 9                          # negate_logical_true.mlir:34:9
	movq	%rbx, %rdi
	callq	_FortranAioEndIoStatement@PLT
	.loc	1 35 3                          # negate_logical_true.mlir:35:3
	addq	$8, %rsp
	.cfi_def_cfa_offset 24
	popq	%rbx
	.cfi_def_cfa_offset 16
	popq	%r14
	.cfi_def_cfa_offset 8
	retq
.Ltmp0:
.Lfunc_end0:
	.size	_QQmain, .Lfunc_end0-_QQmain
	.cfi_endproc
                                        # -- End function
	.type	_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000,@object # @_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000
	.section	.rodata,"a",@progbits
	.weak	_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000
	.p2align	4
_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000:
	.asciz	"./../negate_logical_true.f90"
	.size	_QQcl.2E2F2E2E2F6E65676174655F6C6F676963616C5F747275652E66393000, 29

	.type	_QQcl.78203D20,@object          # @_QQcl.78203D20
	.weak	_QQcl.78203D20
_QQcl.78203D20:
	.ascii	"x = "
	.size	_QQcl.78203D20, 4

	.type	_QQcl.2E4E4F542E78,@object      # @_QQcl.2E4E4F542E78
	.weak	_QQcl.2E4E4F542E78
_QQcl.2E4E4F542E78:
	.ascii	".NOT.x"
	.size	_QQcl.2E4E4F542E78, 6

	.section	.debug_abbrev,"",@progbits
	.byte	1                               # Abbreviation Code
	.byte	17                              # DW_TAG_compile_unit
	.byte	1                               # DW_CHILDREN_yes
	.byte	37                              # DW_AT_producer
	.byte	14                              # DW_FORM_strp
	.byte	19                              # DW_AT_language
	.byte	5                               # DW_FORM_data2
	.byte	3                               # DW_AT_name
	.byte	14                              # DW_FORM_strp
	.byte	16                              # DW_AT_stmt_list
	.byte	23                              # DW_FORM_sec_offset
	.byte	27                              # DW_AT_comp_dir
	.byte	14                              # DW_FORM_strp
	.ascii	"\264B"                         # DW_AT_GNU_pubnames
	.byte	25                              # DW_FORM_flag_present
	.byte	17                              # DW_AT_low_pc
	.byte	1                               # DW_FORM_addr
	.byte	18                              # DW_AT_high_pc
	.byte	6                               # DW_FORM_data4
	.byte	0                               # EOM(1)
	.byte	0                               # EOM(2)
	.byte	2                               # Abbreviation Code
	.byte	46                              # DW_TAG_subprogram
	.byte	0                               # DW_CHILDREN_no
	.byte	17                              # DW_AT_low_pc
	.byte	1                               # DW_FORM_addr
	.byte	18                              # DW_AT_high_pc
	.byte	6                               # DW_FORM_data4
	.byte	64                              # DW_AT_frame_base
	.byte	24                              # DW_FORM_exprloc
	.byte	110                             # DW_AT_linkage_name
	.byte	14                              # DW_FORM_strp
	.byte	3                               # DW_AT_name
	.byte	14                              # DW_FORM_strp
	.byte	58                              # DW_AT_decl_file
	.byte	11                              # DW_FORM_data1
	.byte	59                              # DW_AT_decl_line
	.byte	11                              # DW_FORM_data1
	.byte	63                              # DW_AT_external
	.byte	25                              # DW_FORM_flag_present
	.byte	0                               # EOM(1)
	.byte	0                               # EOM(2)
	.byte	0                               # EOM(3)
	.section	.debug_info,"",@progbits
.Lcu_begin0:
	.long	.Ldebug_info_end0-.Ldebug_info_start0 # Length of Unit
.Ldebug_info_start0:
	.short	4                               # DWARF version number
	.long	.debug_abbrev                   # Offset Into Abbrev. Section
	.byte	8                               # Address Size (in bytes)
	.byte	1                               # Abbrev [1] 0xb:0x39 DW_TAG_compile_unit
	.long	.Linfo_string0                  # DW_AT_producer
	.short	2                               # DW_AT_language
	.long	.Linfo_string1                  # DW_AT_name
	.long	.Lline_table_start0             # DW_AT_stmt_list
	.long	.Linfo_string2                  # DW_AT_comp_dir
                                        # DW_AT_GNU_pubnames
	.quad	.Lfunc_begin0                   # DW_AT_low_pc
	.long	.Lfunc_end0-.Lfunc_begin0       # DW_AT_high_pc
	.byte	2                               # Abbrev [2] 0x2a:0x19 DW_TAG_subprogram
	.quad	.Lfunc_begin0                   # DW_AT_low_pc
	.long	.Lfunc_end0-.Lfunc_begin0       # DW_AT_high_pc
	.byte	1                               # DW_AT_frame_base
	.byte	87
	.long	.Linfo_string3                  # DW_AT_linkage_name
	.long	.Linfo_string3                  # DW_AT_name
	.byte	1                               # DW_AT_decl_file
	.byte	1                               # DW_AT_decl_line
                                        # DW_AT_external
	.byte	0                               # End Of Children Mark
.Ldebug_info_end0:
	.section	.debug_str,"MS",@progbits,1
.Linfo_string0:
	.asciz	"mlir"                          # string offset=0
.Linfo_string1:
	.asciz	"FIRModule"                     # string offset=5
.Linfo_string2:
	.asciz	"/"                             # string offset=15
.Linfo_string3:
	.asciz	"_QQmain"                       # string offset=17
	.section	.debug_pubnames,"",@progbits
	.long	.LpubNames_end0-.LpubNames_start0 # Length of Public Names Info
.LpubNames_start0:
	.short	2                               # DWARF Version
	.long	.Lcu_begin0                     # Offset of Compilation Unit Info
	.long	68                              # Compilation Unit Length
	.long	42                              # DIE offset
	.asciz	"_QQmain"                       # External Name
	.long	0                               # End Mark
.LpubNames_end0:
	.section	.debug_pubtypes,"",@progbits
	.long	.LpubTypes_end0-.LpubTypes_start0 # Length of Public Types Info
.LpubTypes_start0:
	.short	2                               # DWARF Version
	.long	.Lcu_begin0                     # Offset of Compilation Unit Info
	.long	68                              # Compilation Unit Length
	.long	0                               # End Mark
.LpubTypes_end0:
	.section	".note.GNU-stack","",@progbits
	.section	.debug_line,"",@progbits
.Lline_table_start0:
```

**Comment by oroppas** [comment]
I can get the correct result if I build as following

```
bbc -o - ../negate_logical_true.f90 | tco | clang -x ir - -Wno-override-module -L$HOME/packages/llvm/flang/build/lib -Wl,-rpath -Wl,$HOME/packages/llvm/flang/build/lib -lFortran_main -lFortranRuntime -lFortranDecimal
```

**Comment by oroppas** [comment]
I can also get the correct result if I build `flang` with `Debug`.

**Comment by oroppas** [comment]
Simply rebuilding Fortran runtime libraries with Release/Debug can produce wrong/correct result, respectively.

**Comment by jeanPerier** [comment]
> I can also get the correct result if I build flang with Debug.

Thanks a lot for all these logs @oroppas, I was able to reproduce your issue. The very interesting point being that the difference comes from the Fortran runtime library being compiled in debug or release mode.

**Comment by oroppas** [comment]
I'm also curious why compiling through `clang -x ir - ...` is immune to this Fortran runtime library.

**Comment by jeanPerier** [comment]
I believe I may have found the issue. In lowering, we are not adding the `zeroext` attribute in LLVM IR for boolean function argument and returns value. Therefore `.NOT.x` is actually passed  as `0xfe` in the `rsi` register on x86-64 in the call to `_FortranAioOutputLogical`.

This may not play well with the `reinterpret_cast<void *>(&truth)` inside _FortranAioOutputLogical implementation.

I tried modifying the LLVM IR by changing:
 - `declare i1 @_FortranAioOutputLogical(i8*, i1)` to `declare zeroext i1 @_FortranAioOutputLogical(i8*, i1 zeroext )`
 - `%6 = call i1 @_FortranAioOutputLogical(i8* %2, i1 %5)` to `%6 = call zeroext i1 @_FortranAioOutputLogical(i8* %2, i1 zeroext %5)`

It is not entirely clear to me yet if lowering should actually add these `zeroext` attributes, or if the runtime contains an undefined behavior regarding its boolean reinterpret_cast.

**Comment by jeanPerier** [comment]
To complete the analysis, the difference in the runtime between the version compiled with DEBUG and the release version is the code generated by the C++ compiler to store the `truth` boolean argument of _FortranAioOutputLogical [(defined here)](https://github.com/flang-compiler/f18-llvm-project/blob/3ea9d88802029d4231eee032a2d27ab0fa30d413/flang/runtime/io-api.cpp#L1065), so that its address can be passed.

Debug version of the runtime is zero extending the truth argument (in `rsi`) before storing it on the stack:
```
 mov    %sil,%al
 and    $0x1,%al
 mov    %al,-0x9(%rbp)
``` 

The release version is not:
```
  mov    %sil,0xc(%rsp)
```

Which explains the difference of behavior when was not zero extended by the caller `rsi = 0xfe`. The runtime later interpret non all zero bytes as true ([here](https://github.com/flang-compiler/f18-llvm-project/blob/56f1ae387c4453f08e2e6922da51f9ffa59dea4e/flang/runtime/descriptor-io.h#L196)) which explains the difference of behavior between the DEBUG and non DEBUG version of flang on this program.

From a pedantic point of view, I think the fact that the runtime is directly interpreting a boolean storage as an integer with the logic zero is false and the rest is true is slightly wrong because I do not think the C++ standard defines how boolean must be represented in memory. The safe way would probably be to first store the bool argument in an integer (the integer value would be well defined by implicit conversion rules), and then to use the address of this integer that can later be safely reinterpreted as an integer. We may still want to zeroext boolean runtime function arguments in lowering to ba safe though.

**Comment by schweitzpgi** [comment]
I believe this one is fixed. @oroppas Can you confirm on your end?

**Comment by oroppas** [comment]
I can confirm it’s fixed. Thanks!
